src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3152,6 +3153,302 @@ void MacroAssembler::check_klass_subtype(Register sub_klass,
+
+  call_stub(StubRoutines::lookup_secondary_supers_table_slow_path_stub());
+
+  z_bru(L_done); // pass whatever result we got from a slow path
+
+  bind(L_failure);
+  // TODO: use load immediate on condition and z_bru above will not be required
+  z_lghi(r_result, 1);
+
+  bind(L_done);
+  BLOCK_COMMENT("} lookup_secondary_supers_table");
+
+  if (VerifySecondarySupers) {
+    verify_secondary_supers_table(r_sub_klass, r_super_klass, r_result,
+                                  r_temp1, r_temp2, r_temp3);
+  }
+}
+
+// Called by code generated by check_klass_subtype_slow_path
+// above. This is called when there is a collision in the hashed
+// lookup in the secondary supers array.
+void MacroAssembler::lookup_secondary_supers_table_slow_path(Register r_super_klass,
+                                                             Register r_array_base,
+                                                             Register r_array_index,
+                                                             Register r_bitmap,
+                                                             Register r_result,
+                                                             Register r_temp1) {
+  assert_different_registers(r_super_klass, r_array_base, r_array_index, r_bitmap, r_result, r_temp1);
+
+  const Register
+    r_array_length = r_temp1,
+    r_sub_klass    = noreg;
+
+  LOOKUP_SECONDARY_SUPERS_TABLE_REGISTERS;
+
+  BLOCK_COMMENT("lookup_secondary_supers_table_slow_path {");
+  NearLabel L_done, L_failure;
+
+  // Load the array length.
+  z_llgf(r_array_length, Address(r_array_base, Array<Klass*>::length_offset_in_bytes()));
+
+  // And adjust the array base to point to the data.
+  // NB!
+  // Effectively increments the current slot index by 1.
+  assert(Array<Klass*>::base_offset_in_bytes() == wordSize, "");
+  add2reg(r_array_base, Array<Klass*>::base_offset_in_bytes());
+
+  // Linear probe
+  NearLabel L_huge;
+
+  // The bitmap is full to bursting.
+  z_cghi(r_bitmap, Klass::SECONDARY_SUPERS_BITMAP_FULL);
+  z_bre(L_huge);
+
+  // NB! Our caller has checked bits 0 and 1 in the bitmap. The
+  // current slot (at secondary_supers[r_array_index]) has not yet
+  // been inspected, and r_array_index may be out of bounds if we
+  // wrapped around the end of the array.
+
+  { // This is conventional linear probing, but instead of terminating,
+    // when a null entry is found in the table, we maintain a bitmap
+    // in which a 0 indicates missing entries.
+    // The check above guarantees there are 0s in the bitmap, so the loop
+    // eventually terminates.
+
+#ifdef ASSERT
+    // r_result is set to 0 by lookup_secondary_supers_table.
+    // clear_reg(r_result, true /* whole_reg */, false /* set_cc */);
+    z_cghi(r_result, 0);
+    asm_assert(bcondEqual, "r_result required to be 0, used by z_locgr", 44);
+
+    // We should only reach here after having found a bit in the bitmap.
+    // Invariant: array_length == popcount(bitmap)
```

Comment:
It turns out this invariant isn't true. When `array_length` is >= 63 we set `SECONDARY_SUPERS_BITMAP_FULL`, for performance reasons.

```suggestion
```